docs/pages/repo/docs/core-concepts/caching/file-inputs.mdx

```diff
@@ -0,0 +1,116 @@
+
+import Callout from "../../../../../components/Callout";
+
+# File Inputs
+
+By default, a workspace is considered to have been updated when any of the files in that workspace have changed. However, not every file change inside of the entire repo should result in rerunning a task. Turborepo gives you the tools to make sure that you can define every file that matters.
+
+## Global Files
+
+Almost every repository has some configuration that appears at the root of the repository. These files have the potential to impact the behavior of child workspaces.
+
+### `package.json` and `turbo.json`
+
+By default, Turborepo includes the root `package.json`, any lockfile that may be present, and the root `turbo.json` file into the global hash. You do not need to specify them separately.
+
+### Specifying Additional Inputs
+
+If there are additional files that should be considered for every single workspace, such as a root `tsconfig.json` you should specify that in `globalDependencies`:
+
+```jsonc filename="/turbo.json"
+{
+  "$schema": "https://turbo.build/schema.json",
+  "globalDependencies": ["tsconfig.json"],
+  "pipeline": {
+    "type-check": {}
+  }
+}
+```
+
+<Callout type="info">
+  This customization is _additive_ to Turborepo's default behavior for `package.json` and `turbo.json`.
+</Callout>
+
+In this example every single task in the turborepo would be rerun when the root `tsconfig.json` file changes.
+
+## Workspace Files
+
+In addition to global files, which appear outside of any workspace, you may also specify which files to include on a per-workspace basis.
+
+### `package.json` and `turbo.json`
+
+The contents of `package.json` and `turbo.json` are automatically considered for workspace hashes. This behavior is not configurable.
+
+### Turborepo's Default Behavior
+
+The default behavior of Turborepo is to inspect the repository and identify which files should be included in the hash calculation. Turborepo typically uses Git to identify which files to consider:
+
+- All of the files which appear in the Git index which are children of the workspace directory.
+- All of the untracked files which are children of the workspace directory and are not `.gitignore`d.
+
+If either Git or the Git index (typically in `.git/index`) is unavailable, we attempt to approximate the contents which would appear in the Git index using the following strategy:
+
+- Include all of the files which are children of a workspace directory.
+- Exclude all of the files which appear in the root `.gitignore` and the workspace directory's `.gitignore`.
+
+<Callout type="info">
+  These two approaches do not always produce identical results! The scenarios where this does not produce identical results:
+
+- `.gitignore` files at other places in the directory hierarchy.
+- Files added to the Git index which are also specified inside of a `.gitignore`.
+
+You should aim to have both Git and the Git index of your repository available for maximum correctness.
+
+</Callout>
```

Comment:
I would drop this. The fact that they produce different hashes is most likely a bug, and we should fix it.

Specifically the first case will be fixed by switching to the `ignore` crate in Rust, which takes into account the whole hierarchy of `.gitignore` files.

Reply:
Because you can `git add` a `.gitignore`d file it is impossible for the two approaches to be identical. We can resolve the first issue, but not the second.

```
~/repos/git-add-gitignore$ git ls-tree HEAD .
100644 blob 8e8874f882daafd328a3d63da96c5112d62de9cc	.gitignore
100644 blob e69de29bb2d1d6434b8b29ae775ad8c2e48c5391	ignored-file.txt
```

```
~/repos/git-add-gitignore$ cat .gitignore 
ignored-file.txt
```

More generally, I believe that we should document the world as it is, not as we want it to be. This is how we function right now and people need to understand how this works.